### PR TITLE
Fix `init` handling in applyreduce (issue #353)

### LINKED ISSUE #353: `init` being used in applyreduce when I don't think it should
```julia
using GADM
import GeoInterface as GI, GeometryOps as GO
texas = GADM.get("USA", "Texas") |> GI.getfeature |> only |> GI.geometry
texas_lower = GO.applyreduce(vcat, GO.TraitTarget(GI.AbstractCurveTrait), texas) do geom
    points = GO.forcexyz(geom, 0.0).geom
    push!(points, (NaN, NaN, NaN))
    return points
end
```

```julia
114424-element Vector{Union{Nothing, Tuple{Float64, Float64, Float64}}}:
 nothing
 nothing
 (-97.27569599999993, 26.237083000000098, 0.0)
 (-97.27541399999996, 26.237083000000098, 0.0)
 (-97.27541399999996, 26.236528000000135, 0.0)
 (-97.27513899999997, 26.236528000000135, 0.0)
 (-97.27513899999997, 26.235695000000135, 0.0)
 ⋮
 (-97.13500199999987, 28.056389000000024, 0.0)
 (-97.1347199999999, 28.056667000000118, 0.0)
 (-97.13444499999997, 28.056667000000118, 0.0)
 (-97.13416299999989, 28.056944000000158, 0.0)
 (-97.13333099999988, 28.056944000000158, 0.0)
 (NaN, NaN, NaN)
```

That returns a vector where the first 2 elements are nothing, the rest are as expected.  But it's all a bit odd really.  We ought to not put those nothings in.

Passing ` init = (NaN, NaN, NaN)` to applyreduce fixes it for now...but this is the kind of thing we should just do without needing init.

### PR DESCRIPTION
**Pull request by asinghvi17**
## Summary

Fixes #353 - `applyreduce` was incorrectly using `nothing` as an initial value when users didn't provide an explicit `init` keyword, causing spurious `nothing` elements in results when using reducers like `vcat`.

**The problem:** When calling `GO.applyreduce(vcat, target, geom)` without an explicit `init`, the result contained `nothing` values mixed in (e.g., `[nothing, nothing, 4, nothing, 4]` instead of `[4, 4]`).

**The fix:**
- Use `Base._InitialValue()` as a sentinel instead of `nothing` as the default
- Conditionally pass `init` to `mapreduce` only when an actual value was provided
- This matches Julia's standard `reduce`/`mapreduce` semantics

## Changes

- Added `_InitialValue` sentinel type and `_mapreduce_maybe_init` helper function
- Changed `applyreduce` default from `init=nothing` to `init=_InitialValue()`
- Updated all `mapreduce` calls to use the helper
- Added tests for the fix and edge cases
- Updated docstring to clarify `init` behavior

## Test Plan

- [x] New test verifies `vcat` works correctly without explicit `init`
- [x] Edge case tests for empty geometry, explicit init, and numeric reduction
- [x] All existing tests pass
- [x] `area`, `centroid`, `distance` still work correctly

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: GeometryOpsCore/src/applyreduce.jl [hunk 1]
```diff
@@ -3,6 +3,26 @@
 =#
 
 export applyreduce
+public _InitialValue # has to be "public API" since users will have to use it as a default value to init sometimes.
```
Comment (navidcy) [inline GeometryOpsCore/src/applyreduce.jl]: `public` kwarg requires Julia v1.11 but the Project compat is

https://github.com/JuliaGeo/GeometryOps.jl/blob/259c899c0871a45802e22f711a57b09116c4be65/GeometryOpsCore/Project.toml#L17

Am I right?  Perhaps this is related to  buildkite/oceananigans/pr/initialize-enzyme-environment  failing at https://github.com/CliMA/Oceananigans.jl/pull/5147/commits/ce712470bf66647d56ca0239af078f506f95a24e?
